--- ./src/invlib.c.orig	2012-03-06 15:34:39.424814229 -0500
+++ ./src/invlib.c	2012-03-06 15:34:54.787240366 -0500
@@ -228,7 +228,7 @@ invmake(char *invname, char *invpost, FI
 			num = BASE * num + *++s - '!';
 		} while (++i < PRECISION);
 		posting.lineoffset = num;
-		while (++fileindex < nsrcoffset && num > srcoffset[fileindex]) {
+		while (++fileindex < nsrcfiles && num > srcoffset[fileindex]) {
 			;
 		}
 		posting.fileindex = --fileindex;
