--- cscope/src/crossref.c.orig	2003-04-29 10:08:45.000000000 -0400
+++ cscope/src/crossref.c	2004-10-05 10:37:19.958700472 -0400
@@ -95,10 +95,16 @@
 	int	token;			/* current token */
 	struct stat st;
 
+	filename = srcfile;
+	putfilename(srcfile);   /* output the file name */
+	dbputc('\n');
+	dbputc('\n');
+
 	if (! ((stat(srcfile, &st) == 0)
 	       && S_ISREG(st.st_mode))) {
 		cannotopen(srcfile);
 		errorsfound = YES;
+		dbputc('\t');
 		return;
 	}
 	
@@ -107,12 +113,9 @@
 	if ((yyin = myfopen(srcfile, "r")) == NULL) {
 		cannotopen(srcfile);
 		errorsfound = YES;
+		dbputc('\t');
 		return;
 	}
-	filename = srcfile;	/* save the file name for warning messages */
-	putfilename(srcfile);	/* output the file name */
-	dbputc('\n');
-	dbputc('\n');
 
 	/* read the source file */
 	initscanner(srcfile);
