--- cscope-15.5/src/invlib.c.overflow	2003-06-20 03:46:03.000000000 -0400
+++ cscope-15.5/src/invlib.c	2006-04-21 10:28:06.000000000 -0400
@@ -47,7 +47,6 @@
 
 #define	DEBUG		0	/* debugging code and realloc messages */
 #define BLOCKSIZE	2 * BUFSIZ	/* logical block size */
-#define	LINEMAX		1000	/* sorted posting line max size */
 #define	POSTINC		10000	/* posting buffer size increment */
 #define SEP		' '	/* sorted posting field separator */
 #define	SETINC		100	/* posting set size increment */
@@ -112,7 +111,7 @@
 	long	fileindex = 0;	/* initialze, to avoid warning */
 	unsigned postsize = POSTINC * sizeof(POSTING);
 	unsigned long	*intptr;
-	char	line[LINEMAX];
+	char	line[TERMMAX];
 	long	tlong;
 	PARAM	param;
 	POSTING	posting;
@@ -177,7 +176,7 @@
 	lastinblk = sizeof(t_logicalblk);
 
 	/* now loop as long as more to read (till eof)  */
-	while (fgets(line, LINEMAX, infile) != NULL) {
+	while (fgets(line, TERMMAX, infile) != NULL) {
 #if DEBUG || STATS
 		++totpost;
 #endif
